lib/client/db/dbcmd/dbcmd.go

```diff
@@ -226,56 +261,54 @@ func (c *cliCommandBuilder) getMySQLOracleCommand() *exec.Cmd {
 		args = append(args, fmt.Sprintf("--ssl-mode=%s", mysql.MySQLSSLModeVerifyCA))
 	}
 
-	return exec.Command(mysqlBin, args...)
+	return c.exe.Command(mysqlBin, args...)
 }
 
 // getMySQLCommand returns mariadb command if the binary is on the path. Otherwise,
 // mysql command is returned. Both mysql versions (MariaDB and Oracle) are supported.
-func (c *cliCommandBuilder) getMySQLCommand() (*exec.Cmd, error) {
+func (c *CliCommandBuilder) getMySQLCommand() (*exec.Cmd, error) {
 	// Check if mariadb client is available. Prefer it over mysql client even if connecting to MySQL server.
 	if c.isMariaDBBinAvailable() {
 		args := c.getMariaDBArgs()
-		return exec.Command(mariadbBin, args...), nil
-	}
-
-	// Check for mysql binary. Return with error as mysql and mariadb are missing. There is nothing else we can do here.
-	if !c.isMySQLBinAvailable() {
```

Comment:
Is there a way for Teleterm to deal with that error?
1. Currently if none of the supported binaries are installed `tsh` will print `mysql not found not mentioning that `mariadb` is also a supported client
2. If Teleterm fails on every error returned from `CLIBuildder` then another person working on that code can easily return an error and break the Teleterm. If this is the case I'd suggest removing the `error` from returned values.

Reply:
> If Teleterm fails on every error returned from `CLIBuildder` then another person working on that code can easily return an error and break the Teleterm.

Good point.

The fundamental issue is that `tsh db connect` expects `dbcmd` to return a command that will work and if it can't provide one, it should either return an appropriate error or attempt to execute the command which will result in the `<binary> not found` error.

`tsh db connect` also executes `dbcmd` only when the command is actually needed. So if someone has a cluster with a Postgres db but they don't have `psql` installed it and they never try to access it, nothing bad ever happens.

Teleterm on the other hand doesn't expect this command to work. It's more of a "if you have the CLI client installed on your system, this command should connect you to that db". It also needs this command any time it creates a local proxy for the db, but the user might not even want to use that command. They might want to use a GUI app. So again, in this case we don't really care if the binary is actually on the system.

---

@jakule What if we added a `tolerateMissingCLIClient` flag to `connectionCommandOpts`? If it's present, dbcmd would not return an error here, but rather `c.getMySQLOracleCommand()`.

gravitational/teleport#11843 added the `printFormat` flag so that `tsh db config` can wrap Postgres connection strings in quotes. With `tolerateMissingCLIClient`, the option usage would look like this:

use case|printFormat|tolerateMissingCLIClient
-----|-----|-----
tsh db config|✅|❌
tsh db connect|❌|❌
Teleterm|✅|✅

Reply:
@jakule I added a comment which adds said option, please let me know what you think.

Reply:
Fine by me. I'd just change the name of that flag to `ignoreMissingBin` or `skipBinCheck` maybe?

Reply:
Heh, I thought about a name like this. However, for all the other protocols, dbcmd doesn't actually check if the binary is there or not – attempting to run the resulting command is just going to fail with a regular shell error about a missing binary.

At the moment it really is just a special case for MySQL where instead of telling the user that `mysql` doesn't exist we also want to point out that `mariadb` is a valid option too. So I didn't want to use a field name that would suggest that if the field is not present then the check will be done.

In that sense it's more of a caller telling dbcmd "I'm okay with the CLI missing on user's system".